### DISCUSSION TITLE
Claude/openai/google prompt caching

### DISCUSSION THREAD

**Discussion by mrdrprofuroboros** [langfuse/langfuse#2987] [Ideas]
### Describe the feature or potential improvement

Anthropic has recently added prompt caching with its own special pricing
https://www.anthropic.com/news/prompt-caching

Apparently this feature will start popping up in other providers. Unfortunately I didn't find any discussion in langchain either so might worth starting it to get their point of view on the API standard, but right now we have to dirty hack anthropic generations with this sort of workarounds to calculate the price correctly:

```python
anthropic_client = anthropic.AsyncAnthropic()

@observe(as_type="generation")
async def anthropic_completion(model: str, max_tokens: int, system: List[Dict[str, Any]], messages: List[Dict[str, Any]]):
    resp = await anthropic_client.beta.prompt_caching.messages.create(
        model=model,
        max_tokens=max_tokens,
        system=system,
        messages=messages,
    )

    # convert anthropic cache usage to "real" input tokens equivalents
    # https://www.anthropic.com/pricing#anthropic-api
    cache_writes = resp.usage.cache_creation_input_tokens * 3.75 / 3
    cache_reads = resp.usage.cache_read_input_tokens * 0.3 / 3

    langfuse_context.update_current_observation(
        model=model,
        usage={
            "input": resp.usage.input_tokens + cache_writes + cache_reads,
            "output": resp.usage.output_tokens,
        },
    )
    return resp
```

Adding cache writes / reads to model pricing in langfuse would be much cleaner

**Comment by marcklingen** [comment]
Thanks for sharing! We are planning to add more billing logic here as there are some other complexities as well with e.g. google models having different prices depending on the token counts. You can however ingest USD costs if you want to maintain accurate token counts and usd prices in langfuse in the meantime. 

Example ([docs](https://langfuse.com/docs/model-usage-and-cost#cost)):

```python
langfuse_context.update_current_observation(
      usage={
          "input": response.usage.input_tokens,
          "output": response.usage.output_tokens,
          # Optionally, also ingest usd cost. Alternatively, you can infer it via a model definition in Langfuse.
          # Here we assume the input and output cost are 1 USD each.
          "input_cost": 1,
          "output_cost": 1,
          # "total_cost": float, # if not set, it is derived from input_cost + output_cost
      }
  )
```

**Comment by mrdrprofuroboros** [comment]
I made a PR to langchain to support it
https://github.com/langchain-ai/langchain/pull/25644

**Reply by marcklingen** [reply]
awesome, thanks!

**Reply by mrdrprofuroboros** [reply]
but now I'm not able to update the cost in langfuse, cause I can't get into the langchain generation context lol
I made this https://github.com/langfuse/langfuse-python/pull/878
I doubt it's worth merging, but pls take this case into account while working on your billing logic updates

**Comment by antoniomdk** [comment]
OpenAI has recently released [prompt caching](https://openai.com/index/api-prompt-caching/) as well. It'd be great if the Langfuse wrappers could read the cached_tokens from the response and submit the accurate cost.

**Comment by marcklingen** [comment]
This is high up on our roadmap to add, thanks for your +1 @antoniomdk

**Reply by marcklingen** [reply]
cc @hassiebp

**Comment by ngamolsky** [comment]
Any progress on this? Would be great to see it

**Reply by marcklingen** [reply]
@hassiebp has a working PR and will release this soon

**Comment by hassiebp** [comment]
🎄 We have just released support for arbitrary LLM usage types incl. cached / multimodal / reasoning tokens. Feel free to try it out by upgrading your SDKs and adding a custom model definition if necessary, and let us know your feedback!

[Changelog](https://langfuse.com/changelog/2024-12-20-improved-cost-tracking)

**Reply by ajram23** [reply]
@hassiebp what version of the python library supports this? Does the UI show when the LLM call used prompt caching? Lastly is this langchain friendly?

**Reply by hassiebp** [reply]
@ajram23 This is supported for Python SDK versions >2.57. The UI will show when the LLM call used prompt caching, and this change is Langchain-friendly, too. Please let us know whether it works as expected for your chain 🙏🏾

**Reply by ajram23** [reply]
Thanks @hassiebp Do you have a screenshot of how it might look in the UI? Going to try it out today!

**Reply by ajram23** [reply]
@hassiebp any updates on this? I am running 2.57.2 and I have no idea if it's working or not. Can't see any UI differences. Here is an example https://us.cloud.langfuse.com/project/cltolaqjv0000m2pkhvdzv9f6/traces/76fcdd6f-8ab5-44c2-9000-ec92e57137d3?timestamp=2025-01-07T19%3A07%3A57.940Z could you kindly help?

**Reply by hassiebp** [reply]
Hi @ajram23 - thanks a lot for your feedback here and on Discord!

- We have a added a screenshot on getting the usage and cost breakdown by hovering on the respective pill in the generation view to [our docs](https://langfuse.com/docs/model-usage-and-cost)
- We have added sorting by last used model to the models table
- Regarding better highlighting of cached token use directly in the table / generation detail view, I have created this idea post to track demand https://github.com/orgs/langfuse/discussions/4936
- We have added langchain-compatible price keys (`input_cache_read` in your example) such that you will see those cost calculated correctly. Previously, we only supported the official OpenAI / Anthropic keys. If you ingest the same trace again you linked above, the expected behavior should be to see an entry for `input_cache_read` both when hovering on the usage_details and the cost_details

Thanks again for your feedback, AJ - helped a lot as always.

**Reply by ajram23** [reply]
@hassiebp Fantastic as always! Much appreciated! And love the fast iteration!

**Reply by ajram23** [reply]
@hassiebp I see the last used model timestamp. FYI I dont think it's sorted or has the ability to.

**Reply by hassiebp** [reply]
Great! Yes that is expected, we do not yet have shipped custom sorting on the models table. You should see your used models though at the top in descending order by last used 👍🏾

**Reply by ajram23** [reply]
@hassiebp got it. Thats not what I am seeing, for instance I see a whole lot of models I have never used and GPT4-o at the top of the next page. I think you might be right per page but not for the whole list of models.

**Reply by hassiebp** [reply]
Thanks - that's a bug that's fixed here: https://github.com/langfuse/langfuse/pull/4947

**Reply by ajram23** [reply]
@hassiebp works! 🎉

**Comment by mrdrprofuroboros** [comment]
@hassiebp 
I look at the docs https://langfuse.com/docs/model-usage-and-cost
and at the implementation in the repo (though a bit lost there as just skimming fast)

but apparently you're making a bet on OpenAI format, which counts cache reads (great!) but leaves cache writes implicit. Though Anthropic sets a higher price for cache writes (and gives you the control over it)
but your openai-centric model doesn't support anthropic's cache write cost increases, right?

Maybe adding cache_writes field to details and just zeroing it by default would do the trick? It's fine that users would have to provide this manually from claude's responses, it's just the fact that there's no place to put it rn and we still have to do manual shenanigans with input tokens count or cost

**Reply by marcklingen** [reply]
The schema here is flexible, Langfuse supports any keys in the usage_details field and you can add custom price information for it within Langfuse. This should work for cache writes as well

**Reply by marcklingen** [reply]
@hassiebp this might be missing though in the default anthropic prices in langfuse

**Reply by mrdrprofuroboros** [reply]
ah, I see, that's perfect!
might need some more clarity regarding this in the docs, 'cause I only see this feature under the microscope on the UI gif :) (adding arbitrary field)
it is not exposed via API, right? https://api.reference.langfuse.com/#get-/api/public/models

**Reply by marcklingen** [reply]
Thanks! Would really appreciate a contribution to the docs if you have a good idea on how to explain this

> it is not exposed via API, right? https://api.reference.langfuse.com/#get-/api/public/models

Not yet, this will be a follow-up PR, probably needs a v2 api as this is somewhat breaking
Tracking this here: https://github.com/langfuse/langfuse/issues/4811

**Reply by mrdrprofuroboros** [reply]
oof, have to migrate to 3.0 to test it first :D

**Reply by marcklingen** [reply]
Good reason to upgrade, we tried to make this as seamless as possible, let us know if you have any questions

**Reply by hassiebp** [reply]
@mrdrprofuroboros Thanks for your quick feedback :) 

The new cost model allows setting prices for arbitrary usage types for a given model. We have not taken an OpenAI specific approach here: For Anthropic Sonnet 3.5, you may provide your usage details exactly as returned from the Anthropic API, and cost should be accurately computed even when having `cache_read_input_token` or `cache_creation_input_tokens`. [See here](https://github.com/langfuse/langfuse/blob/25b363e1bcc49c472c11cc6d0fcc2b2e0d78afae/worker/src/constants/default-model-prices.json#L1143). 

If you miss a usage type and price combination on a given model, feel free to either open a PR adding the price in `default-model-prices.json` for it to be available to all users or clone the model in the UI and add another key-value pair. 

The only OpenAI-specific utility we provide is to parse the OpenAI Usage schema and map it to Langfuse usage types as described in our docs.
